mixins/trim-whitespace/trim-whitespace-core.js

```diff
@@ -0,0 +1,58 @@
+const isCustomElement = node => node?.tagName?.includes('-');
+
+const addShadowListener = (node, callback) => {
+	if (!node.attachShadow) return;
+	const attachShadow = node.attachShadow.bind(node);
+
+	node.attachShadow = (...args) => {
+		const shadowRoot = attachShadow(...args);
+		callback(shadowRoot);
+		return shadowRoot;
+	};
+};
+
+export class TrimWhitespaceCore {
```

Comment:
Although it's called out in the readme, I'm a little worried about the performance. Some thoughts...

* Consumers could use it liberally... thinking, "isn't is always better to trim this whitespace", which is probably usually true, cost aside.  That became a standard practice back in the Polymer days when they supported `trim-whitespace` on `templates`, though I recall someone from Google saying... "it's performant". 
* It would be possible for an ancestor and a descendant to both be trimming whitespace, resulting in multiple observers observing the same sub-tree. Consumers could end up composing this scenario unknowingly as they use/nest other components that happen to be trimming whitespace.  Maybe there is opportunity for optimization here.
* Do we have a use-case for `deep`? Aside from the same point above applying to an even larger sub-tree, adding more observers, I am also a little worried about modifying other DOM scopes. I know we're just trimming whitespace, but it still doesn't feel right to me.
* This is effectively updating the DOM outside of `render`. Just whitespace, but maybe there is a way to use a directive that doesn't go against that best practice?

Maybe we can find a way to reduce the scope of nodes being observed so that consumers can't apply this to a huge sub-tree?  For example, a directive that only applied to immediate children, or direct children of a slot.

Reply:
So, some answers:

> > directive that only applied to immediate children
>
> I'll investigate whether this is viable 👍.

So, yes it's doable, but it's not great. Usage would look like:

```js
return html`${trimShallow()}|
	<span ${trimShallow()}> ${this.spanText} </span>
	|
	<span ${trimShallow()}> ${this.enableSpanContent ? ' (2 Span Content 2) ' : nothing} </span>
	|
	${this.enableSpanElement ? html`<span ${trimShallow()}> (3 Span Element 3) </span>` : nothing}
	|
	<slot></slot>
|`;
```

Some downsides:

- need to be very verbose with adding the directive to every element
- the directive still needs to use a `MutationObserver`, it just doesn't watch on the full `subtree` in this version
  - unfortunately, the directive `update` method gets called before the DOM is updated, so the trimming can't happen there instead of using a `MutationObserver`

I'll investigate more whether there are other optimizations to be made...